route/grafananet.go

```diff
@@ -125,41 +132,99 @@ func (route *GrafanaNet) run() {
 		}
 	}
 
-	b := &backoff.Backoff{
-		Min:    100 * time.Millisecond,
-		Max:    time.Minute,
-		Factor: 1.5,
-		Jitter: true,
+	go r.run()
+	return r, nil
+}
+
+func (route *GrafanaNet) run() {
+	metrics := make([][]*schema.MetricData, route.concurrency)
+	for i := 0; i < route.concurrency; i++ {
+		metrics[i] = make([]*schema.MetricData, 0, route.flushMaxNum)
+
+		// start up our goroutines for writing data to tsdb-gw
+		route.wg.Add(1)
+		go route.flush(i)
 	}
 
-	flush := func() {
-		if len(metrics) == 0 {
+	flush := func(shard int) {
+		if len(metrics[shard]) == 0 {
 			return
 		}
-
-		mda := schema.MetricDataArray(metrics)
+		mda := schema.MetricDataArray(metrics[shard])
 		data, err := msg.CreateMsg(mda, 0, msg.FormatMetricDataArrayMsgp)
 		if err != nil {
 			panic(err)
 		}
+		route.writeQueue[shard] <- data
+		route.numOut.Inc(int64(len(metrics[shard])))
+		metrics[shard] = metrics[shard][:0]
+
+	}
+
+	hasher := fnv.New32a()
+
+	ticker := time.NewTicker(route.flushMaxWait)
+	for {
+		select {
+		case buf := <-route.buf:
+			route.numBuffered.Dec(1)
+			md := parseMetric(buf, route.schemas)
+			if md == nil {
+				continue
+			}
+			md.SetId()
+			buf = md.KeyBySeries(buf[:0])
```

Comment:
interesting to reuse `buf` here for the key. this is a bit confusing since buf was our metric message, but seems safe since parseMetric returned `md` which is decoupled from any of buf's memory.

i think this warrants a comment though.